feat(learn): Add Tests to Issue Tracker Project, Remove InfoSec Tests

SaintPeter opened the pull request:
Checklist:
- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

Ref: #37185 - Proposed Definition Changes
Ref: #39435 - Remove InfoSoc Tests

Adds all tests to the Issue Tracker Project.  I have tested on my local install against my own project. Explicitly uses the fCC get-tests API to validate that all tests are both passing and have at least one assertion per test.

Note:  These tests assume some of the changes discussed in #39620 are implemented, so the current example project and boilerplate won't pass.  I'll get that pushed out soon.

If you want to grab my code, it's here:
https://github.com/SaintPeter/fcc-issue-tracker-project

SaintPeter committed:
Feature: Add Tests to Issue Tracker Project

raisedadead commented:
This is blocked by #39024

RandellDawson commented:
Fixed merge conflict, and PR #39024 has been merged, so removing `status: blocked` label.

Sky020 commented:
@SaintPeter You are going to need to rebase this to include the test related to the example project.

SaintPeter commented:
@Sky020 
> You are going to need to rebase this to include the test related to the example project.

I'm not sure that I understand.  Do you mean that I need to _add_ a test the validates that they are not submitting the fCC sample project on the `.rocks` domain?  Looking at the code I don't see a test for the sample project at all, so this wouldn't be a _change_.

SaintPeter commented:
Oh, I just realized that the guy who said he was going to make these changes to the boilerplate flaked.  I'll try to figure out what needs doing and do it.  PRs to come soon.

Sky020 commented:
Just to clarify: The test exists. It was just added after this PR was created:
https://github.com/freeCodeCamp/freeCodeCamp/blob/master/curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/issue-tracker.md

So, you need to `git pull --rebase upstream master` to get it into this PR

SaintPeter commented:
@Sky020 There was a user responding to #39620 who said they would make the changes to the boilerplate and they did not.

I will update the boilerplate and sample project to match these updated user stories.  I think I'm going to ease back a bit on the user stories as well, to be more in line with the requirements of the Sudoku Solver and American/British Translator.

While I'm at it, I'll remove the user stories from the `index.html`.

moT01 commented:
Do we want to adjust this to add the new description, and so the test text uses `you`? or should we just keep this as a PR to add tests and make another for the issue of the user stories? I feel like we might as well add it here. What do you think @SaintPeter?

SaintPeter commented:
@moT01 I am planning to rework the boilerplate and sample project to match these updated tests.  I also will update these tests a bit. I should be able to do that in the next few days.

moT01 commented:
I'm fine with that for this one as well, but I think we should try and keep most of the rest of these back end PR's to the scope of the issues - They are difficult to review with PR's to boilerplates and demo-projects and the main repo.

SaintPeter commented:
> They are difficult to review with PR's to boilerplates and demo-projects and the main repo.

@moT01 The issue is that all three are inter-related.  We need the updated tests, because that's the whole point.  We need the boilerplate to match the tests because when users start these projects they need something that is capable of matching and passing the tests on the live site.  Finally, we need the demo projects to pass because that's what we're testing the front-end tests against. It doesn't help that at the same time we're reformatting all of these and moving the user stories around.

I would love to do it all in one PR, but we've got 3 repos for good reasons.

If you have some suggestions on how to make these easier to review, I'm all ears.  Trust me, it's hard enough for me to get them up here and keep them up to date with the underlying repo-wide changes.

SaintPeter committed:
Refactored User Stories

SaintPeter commented:
@nhcarrigan and I met and did a complete revamp of the user stories.  We changed the order, reworded many, and updated some of the requirements to be consistent with the other refactors we've done.  This makes for an overall better challenge.

We additionally rewrote the project description to be more standard and added the instructions block.

These tests can be checked against:
https://issue-tracker.pearlygatesoftware.com/

There is also a PR to update the Boilerplate, here:
https://github.com/freeCodeCamp/boilerplate-project-issuetracker/pull/13

And a PR to update the Demo Project, here:
https://github.com/freeCodeCamp/demo-projects/pull/34


Closes #39620

SaintPeter commented:
Ref: https://github.com/freeCodeCamp/freeCodeCamp/issues/39721

nhcarrigan commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/issue-tracker.md:
```diff
@@ -23,25 +35,242 @@ Start this project on Repl.it using <a href="https://repl.it/github/freeCodeCamp
+        let test_data = {
+          created_by: "fCC",
+        };
+        const data = await $.post(getUserInput("url") + "/api/issues/fcc-project", { created_by: "fCC" });
+        assert.isString(data);
+        assert.property(data, "error");
```
```suggestion
        const data = await $.post(getUserInput("url") + "/api/issues/fcc-project", { created_by: "fCC" });
        assert.isObject(data);
        assert.property(data, "error");
```

This returns an object, not a string.

nhcarrigan commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/issue-tracker.md:
```diff
@@ -23,25 +35,242 @@ Start this project on Repl.it using <a href="https://repl.it/github/freeCodeCamp
+            "_id": itemToUpdate._id, 
+            issue_text: "New Issue Text" 
+          }});
+        asset.isObject(updateSucccess);
```
```suggestion
        assert.isObject(updateSucccess);
```
Typo.

nhcarrigan commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/issue-tracker.md:
```diff
@@ -23,25 +35,242 @@ Start this project on Repl.it using <a href="https://repl.it/github/freeCodeCamp
+        const badUpdate = await $.ajax({url: url, type: "PUT"});
+        assert.isObject(badUpdate);
+        assert.property(badUpdate, "error");
+        assert.equal(badUpdate, "missing _id");
```
```suggestion
        assert.equal(badUpdate.error, "missing _id");
```
Need to assert equality on the property, not the object.

nhcarrigan commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/issue-tracker.md:
```diff
@@ -23,25 +35,242 @@ Start this project on Repl.it using <a href="https://repl.it/github/freeCodeCamp
+    }'
+  - text: If I `PUT` to `/api/issues/{projectname}` and no update fields are sent, return `{ error: 'no update field(s) sent', '_id': _id }`. On any other error, return `{ error: 'could not update ', _id: + _id }`.
+    testString: 'async getUserInput => { 
+      try {
+        const badUpdate = await $.ajax({
+          url: url,
```
```suggestion
      try {
        const url = getUserInput("url") + "/api/issues/fcc-project";
        const badUpdate = await $.ajax({
          url: url,
```
Missing variable declaration here.

nhcarrigan commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/issue-tracker.md:
```diff
@@ -23,25 +35,242 @@ Start this project on Repl.it using <a href="https://repl.it/github/freeCodeCamp
+        throw new Error(err.responseText || err.message);
+      }
+    }'
+  - text: All 11 functional tests are complete and passing.
+    testString: 'async getUserInput => { 
+      try {
+        const getTests = await $.get(getUserInput("url") + "/_api/get-tests" );
+        assert.isArray(getTests);
+        assert.isAtLeast(getTests.length, 11, "At least 11 tests passed");
```
```suggestion
  - text: All 14 functional tests are complete and passing.
    testString: 'async getUserInput => { 
      try {
        const getTests = await $.get(getUserInput("url") + "/_api/get-tests" );
        assert.isArray(getTests);
        assert.isAtLeast(getTests.length, 14, "At least 14 tests passed");
```
We set up 14 tests, not 11.

naomi-lgbt reviewed (changes requested):
This should address the console errors that @RandellDawson was encountering.

SaintPeter committed:
Apply suggestions from code review

Co-authored-by: Nicholas Carrigan (he/him) <nhcarrigan@gmail.com>

SaintPeter commented:
Thanks @nhcarrigan!

Commit:
Modify verbiage for #39721

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

moT01 reviewed (approved):
LGTM 🎉

naomi-lgbt reviewed (approved):
LGTM~! 🥳 Tested locally against both @SaintPeter's live example project, and a locally running version of the demo project PR. Think this is good to go.

Thank you for your pull request! We look forward to additional contributions!
